### PR TITLE
[RELAY][OP] roi_pool operator alter layout

### PR DESCRIPTION
**Pull request by Beya2019**
According to the discussion of roi_align(https://github.com/apache/incubator-tvm/pull/6443)

add aoi_pool operator(in fasterrcnn) convert_op_layout and related test case in test_pass_convert_op_layout.py.

Meantime, I also test the case tests/python/frontend/pytorch/test_object_detection.py(changed the generate_jit_model to fasterrcnn) to verify this roi pool operator.

Would you please have a look at this @kevinthesun @anijain2305. Thanks very much.

### CODE REVIEW

**Comment by Beya2019** [comment]
Hi @kevinthesun:

  can you help to have a look at this submit?  Thanks very much.

File: src/relay/op/vision/rcnn_op.cc [hunk 1]
```diff
@@ -119,14 +119,34 @@ bool ROIPoolRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
...
   CHECK_EQ(rshape.size(), 2) << "Input rois should be 2-D.";
-  CHECK_EQ(roi_pool_attrs->layout, "NCHW") << "ROI Pool only supports NCHW layout";
   // assign output type
-  std::vector<IndexExpr> oshape(
-      {rshape[0], dshape[1], roi_pool_attrs->pooled_size[0], roi_pool_attrs->pooled_size[1]});
+  std::vector<IndexExpr> oshape;
+  if (roi_pool_attrs->layout == "NCHW") {
+    oshape = {rshape[0], dshape[1], roi_pool_attrs->pooled_size[0], roi_pool_attrs->pooled_size[1]};
+  } else {
+    CHECK_EQ(roi_pool_attrs->layout, "NHWC") << "Unexpected ROI Pool layout";
```
Comment (tkonolige) [inline src/relay/op/vision/rcnn_op.cc:127]: Maybe you could add what the expected layout should be?

**Comment by Beya2019** [comment]
Hi @yzhliu, can you help to have a look at this submit? Thanks very much.

File: python/tvm/relay/op/vision/_rcnn.py [hunk 2]
```diff
@@ -69,11 +69,53 @@ def convert_roi_align(attrs, inputs, tinfos, desired_layouts):
...
+    desired_data_layout, desired_rois_layout = map(str, desired_layouts)
+    assert desired_data_layout != "default", "Data layout cannot be default"
+    assert desired_rois_layout == "default", "Rois layout must be default"
+
+    new_attrs["layout"] = desired_data_layout
+    # rois layout not change
+    if desired_data_layout in ["NCHW", "NHWC"]:
+        return relay.vision.roi_pool(data, rois, **new_attrs)
+
+    raise ValueError("Layout %s is not yet supported." % desired_data_layout)
```
Comment (yzhliu) [inline python/tvm/relay/op/vision/_rcnn.py:111]: return None to keep original layout instead of raising exception?

**Review by yzhliu — COMMENTED** [review]
otherwise good to me.

**Review by Beya2019 — COMMENTED** [review]
> otherwise good to me.
Hi @yzhliu :
 
Very Thanks your review, I coded like this for reference conv2d and conv3d:
```
    # Handle default kernel layouts
    if desired_data_layout == "NCHW":
        new_attrs["kernel_layout"] = "OIHW"
        return relay.nn.conv2d(data, weight, **new_attrs)
    elif desired_data_layout == "NHWC":
      ....
        return relay.nn.conv2d(data, weight, **new_attrs)
    raise ValueError("Layout %s is not yet supported." % desired_data_layout)
```
original layout will not go to this process(raising exception)
original layout and Legal desired layout will go the if case which covers the all supported  layout ["NCHW", "NHWC"](As shown below), Only if user gives a not supported layout, the raise ValueError case can report the exception.
```
    if desired_data_layout in ["NCHW", "NHWC"]:
        return relay.vision.roi_pool(data, rois, **new_attrs)

    raise ValueError("Layout %s is not yet supported." % desired_data_layout)
```

**Review by yzhliu — APPROVED** [review]
my bad, was thinking of other stuffs.

**Comment by Beya2019** [comment]
> my bad, was thinking of other stuffs.

Can you help merge the pull request directly? Thanks very much

**Pull request merged**